File: src/main/java/com/example/finsentinel/util/SectorMapper.java
package com.example.finsentinel.util;

/**
 * Provides utility functions for sector mapper operations.
 *
 * <p>This class is part of the util layer in FinSentinel.
 */

public final class SectorMapper {

    /**
     * Creates a new SectorMapper instance.
     *
     * <p>This method is defined in {@link SectorMapper}.
     */

    private SectorMapper() {}

    /**
     * Executes from ticker.
     *
     * <p>This method belongs to {@link SectorMapper} and encapsulates the
     * from ticker workflow.
     * @param ticker ticker (String)
     * @return the from ticker result (String)
     */

    public static String fromTicker(String ticker) {
        if (ticker == null) return null;
        return switch (ticker.toUpperCase()) {
            case "AAPL", "MSFT", "GOOGL", "GOOG", "META", "NVDA", "AMD", "INTC" -> "Technology";
            case "JPM", "BAC", "GS", "MS", "WFC", "C" -> "Financial";
            case "JNJ", "PFE", "UNH", "ABBV", "MRK" -> "Healthcare";
            case "TSLA", "F", "GM" -> "Automotive";
            case "AMZN", "WMT", "TGT", "COST" -> "Retail";
            case "XOM", "CVX", "COP" -> "Energy";
            default -> null;
        };
    }
}

File: packages/shared/src/utils/sector-mapper.ts
/**
 * Ticker-to-sector mapping table.
 *
 * Covers major US equities across all 11 GICS sectors.
 */
const TICKER_SECTOR_MAP: Record<string, string> = {
  // Technology
  AAPL: 'Technology',
  MSFT: 'Technology',
  GOOGL: 'Technology',
  META: 'Technology',
  NVDA: 'Technology',
  AMD: 'Technology',
  INTC: 'Technology',
  TSM: 'Technology',
  AVGO: 'Technology',

  // Financials
  JPM: 'Financials',
  BAC: 'Financials',
  GS: 'Financials',

  // Healthcare
  JNJ: 'Healthcare',
  PFE: 'Healthcare',
  UNH: 'Healthcare',

  // Energy
  XOM: 'Energy',
  CVX: 'Energy',

  // Consumer Discretionary
  AMZN: 'Consumer Discretionary',
  WMT: 'Consumer Discretionary',
  HD: 'Consumer Discretionary',
  TSLA: 'Consumer Discretionary',

  // Consumer Staples
  PG: 'Consumer Staples',
  KO: 'Consumer Staples',
  PEP: 'Consumer Staples',

  // Utilities
  NEE: 'Utilities',
  DUK: 'Utilities',

  // Real Estate
  PLD: 'Real Estate',
  AMT: 'Real Estate',

  // Industrials
  LMT: 'Industrials',
  BA: 'Industrials',

  // Communication Services
  T: 'Communication Services',
  VZ: 'Communication Services',
};

/**
 * Maps a stock ticker symbol to its GICS sector.
 *
 * Case-insensitive lookup. Returns "Unknown" for unrecognized tickers.
 *
 * @param ticker - the stock ticker symbol (e.g. "AAPL", "JPM")
 * @returns the sector name, or "Unknown" if the ticker is not in the map
 */
export function fromTicker(ticker: string): string {
  return TICKER_SECTOR_MAP[ticker.toUpperCase()] ?? 'Unknown';
}
